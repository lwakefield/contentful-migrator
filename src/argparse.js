export default function argparse (args = process.argv.slice(2)) {
    const isLongOpt = v => v && v.startsWith('--')
    const isShortOpt = v => v && v.startsWith('-')
    const isOpt = v => v && isLongOpt(v) || isShortOpt(v)
    const getOptName = v => v && (v.match(/^--?(\w+)/) || [])[1]
    const getOptVal = v => v && (v.match(/^--?\w+=?(.+)?/) || [])[1] || true

    const result = {}
    while (args.length) {
        const arg = args.shift()
        const peek = args[0]

        if (isOpt(arg)) {
            result[getOptName(arg)] = peek && !isOpt(peek)
                ? peek
                : getOptVal(arg)
            if (!isOpt(peek)) args.shift()
        } else {
            result._ = (result._ || []).concat(arg)
        }
    }

    return result
}


