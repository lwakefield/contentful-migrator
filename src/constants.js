import {readFileSync} from 'fs'

export const TEMPLATE_STR = process.env.NODE_ENV === 'build' ?
    require(`raw-loader!${__dirname}/template.js`) :
    readFileSync(`${__dirname}/template.js`).toString()
export const MIGRATIONS_ID = '__migrations'
export const DEFAULT_LOCALE = 'en-US'

