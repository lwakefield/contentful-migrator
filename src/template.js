/**
 * revised_by: null
 * id: <ID>
 * revises: <REVISES>
 */

function up (space) {

}

function down (space) {

}

module.exports = {up, down}
