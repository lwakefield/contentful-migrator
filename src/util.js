import {execSync} from 'child_process'
import {randomBytes} from 'crypto'
import colors from 'colors/safe'

export const ls = path => execSync(`ls ${path}`).toString()
    .split('\n').filter(v => !!v)
export const cp = (from, to, args = '') =>
    execSync(`cp ${args} ${from} ${to} `)
export const rm = (path, args = '') =>
    execSync(`rm ${args} ${path}`)
export const randStr = len => randomBytes(len / 2).toString('hex')

export function log(str, level = 'info') {
    const tags = {
        info: colors.blue('[INFO]'),
        debug: colors.blue('[DEBUG]'),
        error: colors.blue('[ERROR]')
    }
    const tag = tags[level]
    if (!tag) throw new Error(`Invalid log level: ${level}`)

    // eslint-disable-next-line no-console
    console.log(`${tag}: ${str}`)
}

export const info = str => log(str, 'info')
export const debug = str => log(str, 'debug')
export const error = str => log(str, 'error')

export function arrayToObj(arr = [], fn) {
    return arr.reduce(
        (acc, val) => ({...acc, ...fn(val)}),
        {}
    )
}
