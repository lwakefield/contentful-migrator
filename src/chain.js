import {ls, arrayToObj, randStr} from './util'
import {readFileSync, writeFileSync} from 'fs'
import {TEMPLATE_STR} from './constants'

export default class MigrationChain {
    constructor (dir) {
        this.dir = dir
        this.load()
        this.link()
    }
    load () {
        const paths = ls(`${this.dir}/*.js`)
        const migrations = paths.map(v => new Migration(v)).filter(v => !!v.id)

        if (!migrations.length) {
            this.migrations = []
            return
        }

        const migrationMap = arrayToObj(
            migrations,
            (val) => ({[val.id]: val})
        )

        // Backtrack to find the first migration
        let migration = migrations[0]
        while (migration.revisesId) migration = migrationMap[migration.revisesId]

        this.migrations = []
        while (migration) {
            this.migrations.push(migration)
            migration = migrationMap[migration.revisedById]
        }
    }
    link () {
        for (let i = 1; i < this.migrations.length; i++) {
            const [a, b] = [this.migrations[i - 1], this.migrations[i]]
            a.revisedBy = b
            b.revises = a
        }
    }
    createMigration (name) {
        const head = this.migrations[this.migrations.length - 1]
        const headId = head ? head.id : 'null'
        const id = randStr(8)
        const filename = name ?
            `${this.dir}/${id}_${name}.js` :
            `${this.dir}/${id}.js`

        const src = TEMPLATE_STR
            .replace('<ID>', id)
            .replace('<REVISES>', headId)

        writeFileSync(filename, src)

        // Update the old head to reference the new head
        if (head) {
            const headContent = readFileSync(head.path).toString()
            writeFileSync(
                head.path,
                headContent.replace('revised_by: null', `revised_by: ${id}`)
            )
        }

        const migration = new Migration(filename)
        this.migrations.push(migration)
        this.link()
        return migration
    }
    first () {
        return this.migrations[0] || null
    }
    last () {
        return this.migrations[this.migrations.length - 1] || null
    }
    find (id) {
        return this.migrations.find(v => v.id === id) || null
    }
}

export class Migration {
    constructor (path) {
        this.path = path
        this.revises = null
        this.revisedBy = null
        this.load()
    }

    load () {
        const content = readFileSync(this.path).toString()
        const [, idMatch]        = content.match(/\* id: (\w+)/) || []
        const [, revisedByMatch] = content.match(/\* revised_by: (\w+)/) || []
        const [, revisesMatch]   = content.match(/\* revises: (\w+)/) || []

        const isNully = v => !v || v === 'null'

        this.revisedById = isNully(revisedByMatch) ? null: revisedByMatch
        this.id          = isNully(idMatch) ? null: idMatch
        this.revisesId   = isNully(revisesMatch) ? null: revisesMatch
    }

    up (space) {
        return require(this.path).up(space)
    }

    down (space) {
        return require(this.path).down(space)
    }
}
