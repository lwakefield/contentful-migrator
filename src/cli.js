import Chain from './chain'
import Migrator from './migrator'
import {info} from './util'
import argparse from './argparse'

const args = argparse()
const program = (args._ || [])[0]
const {
    CONTENTFUL_SPACE_ID = args['space-id'],
    CONTENTFUL_ACCESS_TOKEN = args['access-token']
} = process.env
process.env.DEBUG = true

export function create () {
    const chain = new Chain(process.cwd())
    const migration = chain.createMigration(args._[1] || '')
    // eslint-disable-next-line no-console
    info(`Created migration ${migration.path}`)
}

export function up () {
    Migrator.get(
        CONTENTFUL_SPACE_ID,
        process.cwd(),
        CONTENTFUL_ACCESS_TOKEN
    ).then(migrator => migrator.upgradeTo(args._[1]))
}

export function down () {
    Migrator.get(
        CONTENTFUL_SPACE_ID,
        process.cwd(),
        CONTENTFUL_ACCESS_TOKEN
    ).then(migrator => migrator.downgradeTo(args._[1]))
}

export function help () {
    /* eslint-disable no-console */
    console.log('usage: contentful-migrator [up|down|create]')
    console.log('  up [ref]:      updates contentful to and including `ref`')
    console.log('                 ref is optional, defaults to the local head')
    console.log('  create [name]: creates a new contentful migration w/ `name`')
    console.log('                 name is optional')
    console.log('  down [ref]:    downgrades contentful to and including ref')
    console.log('                 ref is optional, defaults to the remote head^')
    /* eslint-enable no-console */
}

const toRun = {
    create,
    down,
    help,
    up
}[program] || help

toRun()
