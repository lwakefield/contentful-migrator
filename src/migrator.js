if (!global._babelPolyfill) require('babel-polyfill')

import {ls, randStr, info} from './util'
import {readFileSync, writeFileSync} from 'fs'
import contentful from 'contentful-management'

const TEMPLATE = readFileSync(`${__dirname}/template.js`).toString()
export const MIGRATIONS_ID = '__migrations'
export const DEFAULT_LOCALE = 'en-US'

// TODO: terminology, define the difference b/w a migration and a revision

export const getClient = accessToken => contentful.createClient({accessToken})
export const getSpace = (spaceId, accessToken) =>
    getClient(accessToken).getSpace(spaceId)

export class Migrator {
    constructor (space, dir) {
        this.space = space
        this.dir = dir
        this.migrations = loadRevisionChain(dir)
    }

    // Upgrade to and including
    // TODO: how do you handle a migration failing half way?
    async upgradeTo(revisionId) {
        const remoteHead = getSpaceHead(this.space)
        const firstIndex = this.getMigrationIndex(remoteHead) || -1
        const lastIndex = this.getMigrationIndex(revisionId) ||
            this.migrations.length - 1

        for (let i = firstIndex + 1; i <= lastIndex; i++) {
            const id = this.migrations[i].id
            const migration = this.loadMigration(id)
            console.log(`running upgrade: ${id}`)
            await migration.up(this.space)
            await this.updateRemoteHead(id)
        }
    }

    async downgradeTo (revisionId) {
        const remoteHead = await getSpaceHead(this.space)
        if (!remoteHead) throw new Error('Could not get the remote head')

        const firstIndex = this.getMigrationIndex(remoteHead)
        const lastIndex = this.getMigrationIndex(revisionId) || 0

        for (let i = firstIndex; i >= lastIndex; i--) {
            const id = this.migrations[i].id
            const migration = this.loadMigration(id)
            console.log(`running downgrade: ${id}`)
            await migration.down(this.space)
            await this.updateRemoteHead(id)
        }
    }

    loadMigration(id) {
        const migration = this.migrations.find(v => v.id === id)
        if (!migration) throw new Error('Migration not found')

        return require(migration.path)
    }

    updateRemoteHead (id) {
        return this.space.createEntry(
            MIGRATIONS_ID,
            {fields: {ref: {[DEFAULT_LOCALE]: id}}}
        )
    }

    getMigration (id) {
        return this.migrations.find(v => v.id === id) || null
    }

    getMigrationIndex (id) {
        return this.migrations.findIndex(v => v.id === id) || null
    }
}

export async function paginate(fn, query = {}) {
    const limit = 1000
    const entries = []
    let isDone = false
    while (!isDone) {
        const page = await fn({...query, skip: entries.length, limit})
        const items = page.items || []
        entries.push(...items)
        isDone = !items.length
    }
    return entries
}

export async function getSpaceHead(space) {
    const page = await space.getEntries({
        content_type: MIGRATIONS_ID,
        order: '-sys.createdAt',
        limit: 1
    })
    const items = (page.items || [])
    return items.length ?
        items[0].fields.ref[DEFAULT_LOCALE] :
        null
}

export async function getMigrations (space) {
    return await paginate(
        space.getEntries,
        {content_type: MIGRATIONS_ID, order: 'sys.createdAt'}
    )
}

export async function prepareSpace (space) {
    try {
        const migrations = await space.getContentType(MIGRATIONS_ID)
        info(`${MIGRATIONS_ID} content type already exists`)
        return migrations
    } catch (e) {}

    info(`${MIGRATIONS_ID} content type does not exist.`)
    info(`creating ${MIGRATIONS_ID} content type`)


    return await space.createContentTypeWithId(
        MIGRATIONS_ID,
        {name: 'Migrations', fields: [
            {name: 'ref', id: 'ref', type: 'Symbol'}
        ]}
    ).then(v => v.publish())
}

export function createMigration (path, name = '') {
    const revisionChain = loadRevisionChain(path)
    const head = revisionChain[revisionChain.length - 1]
    const id = randStr(8)
    const src = TEMPLATE
        .replace('<ID>', id)
        .replace('<REVISES>', head.id)

    const filename = name ?
        `${path}/${id}_${name}.js` :
        `${path}/${id}.js`

    writeFileSync(filename, src)
    return filename
}

export function loadRevisionChain (dir) {
    const paths = ls(`${dir}/*.js`)

    const revisionHash = {}
    for (const path of paths) {
        const content = readFileSync(path).toString()
        let [, revised_by] = content.match(/\* revised_by: (\w+)/)
        const [, id] = content.match(/\* id: (\w+)/)
        let [, revises] = content.match(/\* revises: (\w+)/)
        if (revised_by === 'null') revised_by = null
        if (revises === 'null') revises = null

        revisionHash[id] = {revised_by, id, revises, path}
    }

    // TODO: this could do with some tidying to make sure we don't leave any
    // hanging revisions...

    let node = revisionHash[Object.keys(revisionHash)[0]]
    while (node.revises) node = revisionHash[node.revises]

    let revisionChain = []

    while (node) {
        revisionChain.push(node)
        node = revisionHash[node.revised_by]
    }

    return revisionChain
}
