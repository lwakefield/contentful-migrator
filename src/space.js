import contentful from 'contentful-management'

import {
    MIGRATIONS_ID,
    DEFAULT_LOCALE
} from './constants'
import {info} from './util'

export default class Space {
    constructor (space) {
        this.space = space
        this.migrations = []
    }
    static async get (id, accessToken) {
        const _space = await contentful.createClient({accessToken}).getSpace(id)
        const space = new Space(_space)

        try {
            await space.loadMigrationHistory()
        } catch (e) {
            info(`${MIGRATIONS_ID} content type does not exist`)
            info(`creating ${MIGRATIONS_ID} content type`)
            await space.initSpace()
        }

        return space
    }

    async loadMigrationHistory () {
        this.migrations = await enumerate(
            this.space.getEntries,
            {content_type: MIGRATIONS_ID, order: 'sys.createdAt'}
        )
    }

    async initSpace () {
        const contentType = await this.space.createContentTypeWithId(
            MIGRATIONS_ID,
            {name: 'Migrations', fields: [
                {name: 'ref', id: 'ref', type: 'Symbol'}
            ]}
        )
        await contentType.publish()
    }

    async addHead (id) {
        const migration = await this.space.createEntry(
            MIGRATIONS_ID,
            {fields: {ref: {[DEFAULT_LOCALE]: id}}}
        )
        this.migrations.push(migration)
    }

    async deleteHead () {
        const head = this.migrations.shift()
        await head.delete()
    }

    getHeadRef () {
        const last = this.migrations[this.migrations.length - 1]
        return last ? last.fields.ref[DEFAULT_LOCALE] : null
    }
}

export async function enumerate (fn, query = {}) {
    const limit = 1000
    const entries = []
    let isDone = false
    while (!isDone) {
        const page = await fn({...query, skip: entries.length, limit})
        const items = page.items || []
        entries.push(...items)
        isDone = !items.length
    }
    return entries
}
