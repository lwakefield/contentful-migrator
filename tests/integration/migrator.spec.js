require('dotenv').config()
jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;

import {
    MIGRATIONS_ID,
    DEFAULT_LOCALE,
    getSpace,
    getSpaceHead,
    paginate,
    Migrator
} from '../../src/migrator'
import {ls, cp, rm} from '../../src/util'
import {tmpdir} from 'os'
import fs from 'fs'

const FIXTURE_PATH = `${__dirname}/../fixtures`
const TEMPLATE = fs.readFileSync(`${__dirname}/../../src/template.js`).toString()

let space
beforeAll(async () => {
    space = await getSpace(
        process.env.CONTENTFUL_SPACE_ID,
        process.env.CONTENTFUL_ACCESS_TOKEN
    )
})

test('getSpace', async () => {
    const space = await getSpace(
        process.env.CONTENTFUL_SPACE_ID,
        process.env.CONTENTFUL_ACCESS_TOKEN
    )
    expect(space).toBeTruthy()
})

// test('getSpaceHead', async () => {
//     const headEntry = await space.createEntry(MIGRATIONS_ID, {
//         fields: {ref: {[DEFAULT_LOCALE]: 'abc123'}}
//     })
//     expect(headEntry).toBeTruthy()
//     const head = await getSpaceHead(space)
//     expect(head).toEqual('abc123')
//     await headEntry.delete()
// })

describe('Migrator', () => {
    it('instantiates correctly', () => {
        const migrator = new Migrator(space, FIXTURE_PATH)
        expect(migrator).toBeTruthy()
        expect(migrator.migrations).toMatchSnapshot()
    })
    it('loads migrations', () => {
        const migrator = new Migrator(space, FIXTURE_PATH)
        const firstMigration = migrator.migrations[0]
        const migration = migrator.loadMigration(firstMigration.id)
        expect(migration).toMatchSnapshot()
    })
    it('migrations loaded runs up and down correctly', async () => {
        const migrator = new Migrator(space, FIXTURE_PATH)
        const firstMigration = migrator.migrations[0]
        const migration = migrator.loadMigration(firstMigration.id)

        const up = await migration.up(space)
        expect(up.sys).toBeTruthy()
        expect(up.name).toMatchSnapshot()
        expect(up.fields).toMatchSnapshot()

        const down = await migration.down(space)
        expect(down).toEqual(undefined)
    })
    it('runs an upgrade with two migrations', async () => {
        const migrator = new Migrator(space, FIXTURE_PATH)
        const migration = migrator.migrations[1].id

        // TODO we need some real assertions in here...
        await migrator.upgradeTo(migration)

        await new Promise(res => setTimeout(res, 2000))

        await migrator.downgradeTo()
    })
})
