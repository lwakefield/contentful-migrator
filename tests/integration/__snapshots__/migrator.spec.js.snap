exports[`Migrator instantiates correctly 1`] = `
Array [
  Object {
    "id": "907230f7c",
    "path": "/Users/lawrence/projects/contentful-migrator/tests/integration/../fixtures/907230f7c.js",
    "revised_by": "55da42387",
    "revises": null,
  },
  Object {
    "id": "55da42387",
    "path": "/Users/lawrence/projects/contentful-migrator/tests/integration/../fixtures/55da42387.js",
    "revised_by": "c401e90d5",
    "revises": "907230f7c",
  },
  Object {
    "id": "c401e90d5",
    "path": "/Users/lawrence/projects/contentful-migrator/tests/integration/../fixtures/c401e90d5.js",
    "revised_by": null,
    "revises": "55da42387",
  },
]
`;

exports[`Migrator loads migrations 1`] = `
Object {
  "down": [Function],
  "up": [Function],
}
`;

exports[`Migrator migrations loaded runs up and down correctly 1`] = `"testContentType"`;

exports[`Migrator migrations loaded runs up and down correctly 2`] = `
Array [
  Object {
    "disabled": false,
    "id": "testField",
    "localized": false,
    "name": "testField",
    "omitted": false,
    "required": false,
    "type": "Symbol",
    "validations": Array [],
  },
]
`;
